feature: add more integation test case for qos sort plugin

googs1025 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
- add more integation test case for qos sort plugin
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #None

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
None
```

googs1025 commented:
/test pull-scheduler-plugins-integration-test-master

googs1025 commented:
/test pull-scheduler-plugins-integration-test-master

ffromani commented:
/cc

ffromani commented on test/integration/qos_test.go:
```diff
@@ -171,3 +171,112 @@ func TestQOSPlugin(t *testing.T) {
+	syncInformerFactory(testCtx)
+	defer cleanupTest(t, testCtx)
+
+	ns := fmt.Sprintf("integration-test-%v", string(uuid.NewUUID()))
```
nit: can we put `qos` anywhere in the name, to make it obvious which test belong to which suite a runtime?
it's a cheap change which can improve the quality of life in troubleshooting

googs1025 replied:
After thinking about this, maybe we can merge the above `TestQOSPlugin` and use the following below, which may be better.

```go
func TestQOSSuite(t *testing.T) {
    t.Run("DifferentQoS", func(t *testing.T) {
        testCtx := &testContext{}
        testCtx.Ctx, testCtx.CancelFn = context.WithCancel(context.Background())
        
        ...
    }
    
    t.Run("SameQoSDifferentCreationTime", func(t *testing.T) {
        testCtx := &testContext{}
        testCtx.Ctx, testCtx.CancelFn = context.WithCancel(context.Background())
        
        ...
    }
}

```

ffromani replied:
I don't have strong opinions, both would work for me

ffromani commented on test/integration/qos_test.go:
```diff
@@ -171,3 +171,112 @@ func TestQOSPlugin(t *testing.T) {
+	ns := fmt.Sprintf("integration-test-%v", string(uuid.NewUUID()))
+	createNamespace(t, testCtx, ns)
+
+	// Create a Node.
+	nodeName := "fake-node"
+	node := st.MakeNode().Name(nodeName).Label("node", nodeName).Obj()
+	node.Status.Capacity = v1.ResourceList{
+		v1.ResourcePods:   *resource.NewQuantity(32, resource.DecimalSI),
+		v1.ResourceCPU:    *resource.NewMilliQuantity(500, resource.DecimalSI),
+		v1.ResourceMemory: *resource.NewQuantity(500, resource.DecimalSI),
+	}
```
nit: can you add a comment to document why we use these settings? perhaps `we don't really need too many resources, so pick semi-random "low enough" values` (note this is pure guesswork of mine)

ffromani commented on test/integration/qos_test.go:
```diff
@@ -171,3 +171,112 @@ func TestQOSPlugin(t *testing.T) {
+
+	// Wait for all Pods are in the scheduling queue.
+	err = wait.PollUntilContextTimeout(testCtx.Ctx, time.Millisecond*200, wait.ForeverTestTimeout, false, func(ctx context.Context) (bool, error) {
+		pendingPods, _ := testCtx.Scheduler.SchedulingQueue.PendingPods()
```
question: what guarantees we will find the pods in the pending pods at time of checking?
IOW, can this ever be racy?

ffromani replied:
ok, makes sense. I need to refresh my memory of what I did in the noderesourcetopology tests but IIRC I called `Run` and here you don't, and this should explain everything

ffromani reviewed (commented):
LGTM with some questions inside
minor nits, please address them if you have to resubmit anyway

Huang-Wei commented:
/hold to confirm if this PR is really needed.

googs1025 commented on test/integration/qos_test.go:
```diff
@@ -42,132 +42,254 @@ import (
-	nodeName := "fake-node"
-	node := st.MakeNode().Name("fake-node").Label("node", nodeName).Obj()
-	node.Status.Capacity = v1.ResourceList{
-		v1.ResourcePods:   *resource.NewQuantity(32, resource.DecimalSI),
```
remove

googs1025 commented on test/integration/qos_test.go:
```diff
@@ -42,132 +42,254 @@ import (
+		// - Maximum 6 Pods: Since we only create a total of 6 Pods across two test cases, this number is sufficient.
+		// - 500 millicores (0.5 cores) CPU: This exceeds the combined CPU requests of all test Pods (200 millicores), providing a buffer.
+		// - 500 MiB Memory: This also exceeds the combined memory requests of all test Pods (300 MiB), ensuring there's enough memory available.
+		// Note that these settings are based on the current test requirements and may need adjustment if the tests change in the future.
```
removed

googs1025 commented:
> /hold to confirm if this PR is really needed.

I think this PR is reasonable. We added qos class sorting and pod creation time in QosSortPlugin in this https://github.com/kubernetes-sigs/scheduler-plugins/pull/855 In the original use case, we only had different qos. Now, we added the use case with the same qos but different pod creation time.

googs1025 commented on test/integration/qos_test.go:
```diff
@@ -42,132 +42,242 @@ import (
+		t.Logf("Start to create 3 Guaranteed Pods sequentially.")
+		for _, pod := range pods {
+			// Delay between creations to ensure different timestamps.
+			// time.Sleep(time.Second * time.Duration(i))
```
for testing, if no unstable flaky failed, I will remove this

googs1025 replied:
https://prow.k8s.io/pr-history/?org=kubernetes-sigs&repo=scheduler-plugins&pr=866 
it seem to no unstable flaky

googs1025 replied:
already removed

googs1025 commented:
/test pull-scheduler-plugins-integration-test-master


for testing ci

googs1025 commented:
/test pull-scheduler-plugins-integration-test-master

googs1025 commented:
/test pull-scheduler-plugins-integration-test-master

googs1025 commented:
ping @Huang-Wei :)

Huang-Wei reviewed (commented):
/lgtm
/approve

googs1025 commented:
/hold cancel
